File: apps/nomadically.work/langgraph/src/graphs/tech_knowledge/taxonomy.py
"""Maps skill tags to knowledge categories with hierarchy definitions."""

from src.models.taxonomy import SKILL_TAGS, SKILL_LABELS

# ── Category definitions for the knowledge app ────────────────────────
# Each category: (name, icon, description, gradient_from, gradient_to)

TECH_CATEGORIES: dict[str, dict] = {
    "Databases & Storage": {
        "icon": "\U0001f5c4\ufe0f",
        "description": "Database systems, caching, and data storage technologies",
        "gradient_from": "#0891b2",
        "gradient_to": "#22d3ee",
    },
    "Backend Frameworks": {
        "icon": "\u2699\ufe0f",
        "description": "Server-side frameworks and runtime environments",
        "gradient_from": "#059669",
        "gradient_to": "#34d399",
    },
    "Frontend Frameworks": {
        "icon": "\U0001f3a8",
        "description": "Client-side frameworks, libraries, and UI toolkits",
        "gradient_from": "#7c3aed",
        "gradient_to": "#a78bfa",
    },
    "Cloud & DevOps": {
        "icon": "\u2601\ufe0f",
        "description": "Cloud platforms, containers, orchestration, and CI/CD",
        "gradient_from": "#2563eb",
        "gradient_to": "#60a5fa",
    },
    "Languages": {
        "icon": "\U0001f4dd",
        "description": "Programming languages and their ecosystems",
        "gradient_from": "#d97706",
        "gradient_to": "#fbbf24",
    },
    "Testing & Quality": {
        "icon": "\u2705",
        "description": "Testing frameworks, quality assurance, and development practices",
        "gradient_from": "#e11d48",
        "gradient_to": "#fb7185",
    },
    "API & Communication": {
        "icon": "\U0001f50c",
        "description": "API protocols, real-time communication, and integration patterns",
        "gradient_from": "#6366f1",
        "gradient_to": "#818cf8",
    },
    "LLM Frameworks": {
        "icon": "\U0001f9e0",
        "description": "LLM orchestration frameworks, agents, and AI application toolkits",
        "gradient_from": "#9333ea",
        "gradient_to": "#c084fc",
    },
}

# ── Tag to category mapping ───────────────────────────────────────────

TAG_TO_CATEGORY: dict[str, str] = {
    # Databases & Storage
    "postgresql": "Databases & Storage",
    "mysql": "Databases & Storage",
    "mongodb": "Databases & Storage",
    "redis": "Databases & Storage",
    "elasticsearch": "Databases & Storage",
    "cassandra": "Databases & Storage",
    "dynamodb": "Databases & Storage",
    "sqlite": "Databases & Storage",
    "sql": "Databases & Storage",
    "vector-db": "Databases & Storage",
    "pinecone": "Databases & Storage",
    "weaviate": "Databases & Storage",
    "chromadb": "Databases & Storage",
    "qdrant": "Databases & Storage",
    "pgvector": "Databases & Storage",
    "cloudflare-d1": "Databases & Storage",

    # Backend Frameworks
    "nodejs": "Backend Frameworks",
    "express": "Backend Frameworks",
    "django": "Backend Frameworks",
    "flask": "Backend Frameworks",
    "laravel": "Backend Frameworks",
    "fastapi": "Backend Frameworks",
    "spring-boot": "Backend Frameworks",
    "hono": "Backend Frameworks",
    "trpc": "Backend Frameworks",
    "drizzle-orm": "Backend Frameworks",
    "prisma": "Backend Frameworks",
    "langchain": "LLM Frameworks",
    "langgraph": "LLM Frameworks",
    "llamaindex": "LLM Frameworks",

    # ML & AI (LLM Frameworks category)
    "pytorch": "LLM Frameworks",
    "tensorflow": "LLM Frameworks",
    "huggingface": "LLM Frameworks",
    "openai": "LLM Frameworks",
    "anthropic": "LLM Frameworks",
    "vercel-ai-sdk": "LLM Frameworks",
    "mastra": "LLM Frameworks",
    "machine-learning": "LLM Frameworks",
    "deep-learning": "LLM Frameworks",
    "llm": "LLM Frameworks",
    "rag": "LLM Frameworks",
    "prompt-engineering": "LLM Frameworks",
    "fine-tuning": "LLM Frameworks",
    "embeddings": "LLM Frameworks",
    "transformers": "LLM Frameworks",
    "agents": "LLM Frameworks",
    "agentic-ai": "LLM Frameworks",
    "mlops": "LLM Frameworks",
    "model-evaluation": "LLM Frameworks",
    "structured-output": "LLM Frameworks",
    "function-calling": "LLM Frameworks",
    "nlp": "LLM Frameworks",
    "computer-vision": "LLM Frameworks",
    "langsmith": "LLM Frameworks",
    "langfuse": "LLM Frameworks",
    "deepeval": "Testing & Quality",
    "sagemaker": "Cloud & DevOps",
    "vertex-ai": "Cloud & DevOps",
    "bedrock": "Cloud & DevOps",

    # ML data tools (Backend Frameworks — computation libraries)
    "pandas": "Backend Frameworks",
    "numpy": "Backend Frameworks",
    "scikit": "Backend Frameworks",

    # Frontend Frameworks
    "react": "Frontend Frameworks",
    "vue": "Frontend Frameworks",
    "angular": "Frontend Frameworks",
    "svelte": "Frontend Frameworks",
    "nextjs": "Frontend Frameworks",
    "remix": "Frontend Frameworks",
    "astro": "Frontend Frameworks",
    "react-native": "Frontend Frameworks",
    "flutter": "Frontend Frameworks",
    "tailwind": "Frontend Frameworks",
    "radix-ui": "Frontend Frameworks",
    "shadcn-ui": "Frontend Frameworks",
    "react-query": "Frontend Frameworks",
    "zustand": "Frontend Frameworks",
    "apollo-client": "Frontend Frameworks",

    # Cloud & DevOps
    "aws": "Cloud & DevOps",
    "gcp": "Cloud & DevOps",
    "azure": "Cloud & DevOps",
    "docker": "Cloud & DevOps",
    "kubernetes": "Cloud & DevOps",
    "terraform": "Cloud & DevOps",
    "ansible": "Cloud & DevOps",
    "jenkins": "Cloud & DevOps",
    "ci-cd": "Cloud & DevOps",
    "circleci": "Cloud & DevOps",
    "serverless": "Cloud & DevOps",
    "microservices": "Cloud & DevOps",
    "cloudflare-workers": "Cloud & DevOps",
    "cloudflare-workers-ai": "Cloud & DevOps",
    "cloudflare-vectorize": "Cloud & DevOps",
    "linux": "Cloud & DevOps",
    "git": "Cloud & DevOps",

    # Languages
    "javascript": "Languages",
    "typescript": "Languages",
    "python": "Languages",
    "java": "Languages",
    "csharp": "Languages",
    "ruby": "Languages",
    "php": "Languages",
    "go": "Languages",
    "rust": "Languages",
    "swift": "Languages",
    "kotlin": "Languages",
    "scala": "Languages",
    "elixir": "Languages",
    "bun": "Languages",
    "deno": "Languages",

    # Testing & Quality
    "jest": "Testing & Quality",
    "pytest": "Testing & Quality",
    "playwright": "Testing & Quality",
    "cypress": "Testing & Quality",
    "vitest": "Testing & Quality",
    "storybook": "Testing & Quality",
    "tdd": "Testing & Quality",
    "agile": "Testing & Quality",
    "webpack": "Testing & Quality",

    # API & Communication
    "rest-api": "API & Communication",
    "graphql": "API & Communication",
    "grpc": "API & Communication",
    "websocket": "API & Communication",
    "event-driven": "API & Communication",
}


def get_category_for_tag(tag: str) -> str:
    """Return the knowledge category for a skill tag, or 'Languages' as fallback."""
    return TAG_TO_CATEGORY.get(tag, "Languages")


def get_label_for_tag(tag: str) -> str:
    """Return the human-readable label for a skill tag."""
    return SKILL_LABELS.get(tag, tag.replace("-", " ").title())


def normalize_tag(raw: str) -> str | None:
    """Normalize a raw technology name to a canonical skill tag, or None if unknown."""
    lowered = raw.lower().strip().replace(" ", "-").replace(".", "").replace("/", "-")

    # Direct match
    if lowered in SKILL_TAGS:
        return lowered

    # Common aliases
    aliases: dict[str, str] = {
        "postgres": "postgresql",
        "pg": "postgresql",
        "node": "nodejs",
        "node-js": "nodejs",
        "express-js": "express",
        "expressjs": "express",
        "next-js": "nextjs",
        "next": "nextjs",
        "react-js": "react",
        "reactjs": "react",
        "vue-js": "vue",
        "vuejs": "vue",
        "angular-js": "angular",
        "angularjs": "angular",
        "svelte-js": "svelte",
        "sveltejs": "svelte",
        "tailwindcss": "tailwind",
        "tailwind-css": "tailwind",
        "k8s": "kubernetes",
        "tf": "terraform",
        "mongo": "mongodb",
        "elastic": "elasticsearch",
        "es": "elasticsearch",
        "dynamo": "dynamodb",
        "amazon-web-services": "aws",
        "google-cloud": "gcp",
        "google-cloud-platform": "gcp",
        "microsoft-azure": "azure",
        "c-sharp": "csharp",
        "c#": "csharp",
        "spring": "spring-boot",
        "springboot": "spring-boot",
        "fast-api": "fastapi",
        "scikit-learn": "scikit",
        "sklearn": "scikit",
        "rlhf": "rlhf-preference",
        "ci": "ci-cd",
        "cd": "ci-cd",
        "cicd": "ci-cd",
        "rest": "rest-api",
        "restful": "rest-api",
        "llama-index": "llamaindex",
        "llama_index": "llamaindex",
        # ML / AI aliases
        "hugging-face": "huggingface",
        "hugging_face": "huggingface",
        "pg-vector": "pgvector",
        "pg_vector": "pgvector",
        "lang-smith": "langsmith",
        "lang-fuse": "langfuse",
        "vertexai": "vertex-ai",
        "vertex_ai": "vertex-ai",
        "google-vertex-ai": "vertex-ai",
        "google-vertex": "vertex-ai",
        "aws-sagemaker": "sagemaker",
        "amazon-sagemaker": "sagemaker",
        "aws-bedrock": "bedrock",
        "amazon-bedrock": "bedrock",
        "open-ai": "openai",
        "claude": "anthropic",
        "gemini": "gcp",
    }
    if lowered in aliases:
        return aliases[lowered]

    return None


def make_cat_slug(cat_name: str) -> str:
    """Convert a category display name to a URL slug."""
    return cat_name.lower().replace(" & ", "-").replace(" ", "-")


def make_lesson_slug(tag: str) -> str:
    """Build a hierarchical lesson slug: {category-slug}/{tag}."""
    cat_name = get_category_for_tag(tag)
    return f"{make_cat_slug(cat_name)}/{tag}"

File: apps/knowledge/ml/topic-miner/src/taxonomy.rs
//! Skill taxonomy — ported from lead-gen's skill-taxonomy.ts.
//!
//! Each entry maps a kebab-case slug to a human label, category, and a set of
//! regex patterns that detect the skill in source code.

use std::collections::HashMap;
use std::sync::LazyLock;

#[derive(Debug, Clone)]
pub struct SkillDef {
    pub tag: &'static str,
    pub label: &'static str,
    pub category: &'static str,
    /// Regex patterns (case-insensitive) that signal this skill in code.
    pub patterns: &'static [&'static str],
}

/// All canonical skill tags with detection patterns.
pub static TAXONOMY: LazyLock<HashMap<&'static str, SkillDef>> = LazyLock::new(|| {
    let entries: Vec<SkillDef> = vec![
        // ── Programming Languages ────────────────────────────────────────────
        s("javascript", "JavaScript", "Languages", &[r#"\.js$"#, r#"\.mjs$"#, r#"\.cjs$"#]),
        s("typescript", "TypeScript", "Languages", &[r#"\.ts$"#, r#"\.tsx$"#, "tsconfig"]),
        s("python", "Python", "Languages", &[r#"\.py$"#, r#"import\s+\w+"#, r#"from\s+\w+\s+import"#]),
        s("java", "Java", "Languages", &[r#"\.java$"#, r#"import\s+java\."#]),
        s("csharp", "C#", "Languages", &[r#"\.cs$"#, r#"using\s+System"#]),
        s("ruby", "Ruby", "Languages", &[r#"\.rb$"#, r#"require\s+['"]\w+"#]),
        s("php", "PHP", "Languages", &[r#"\.php$"#, r#"<\?php"#]),
        s("go", "Go", "Languages", &[r#"\.go$"#, r#"package\s+main"#, r#"import\s+\("#]),
        s("rust", "Rust", "Languages", &[r#"\.rs$"#, r#"Cargo\.toml"#, r#"use\s+std::"#]),
        s("swift", "Swift", "Languages", &[r#"\.swift$"#, r#"import\s+Foundation"#]),
        s("kotlin", "Kotlin", "Languages", &[r#"\.kt$"#, r#"import\s+kotlin"#]),
        s("scala", "Scala", "Languages", &[r#"\.scala$"#]),
        s("elixir", "Elixir", "Languages", &[r#"\.ex$"#, r#"\.exs$"#, "defmodule"]),

        // ── Frontend Frameworks ──────────────────────────────────────────────
        s("react", "React", "Frontend", &[r#"from\s+['"]react['"]"#, r#"useState|useEffect|useRef|useCallback"#]),
        s("vue", "Vue.js", "Frontend", &[r#"from\s+['"]vue['"]"#, r#"\.vue$"#]),
        s("angular", "Angular", "Frontend", &[r#"@angular/"#, r#"\.component\.ts$"#]),
        s("svelte", "Svelte", "Frontend", &[r#"\.svelte$"#, r#"from\s+['"]svelte['"]"#]),
        s("nextjs", "Next.js", "Frontend", &[r#"from\s+['"]next"#, r#"next\.config"#, r#"app/layout\."#]),

        // ── Backend Frameworks ───────────────────────────────────────────────
        s("nodejs", "Node.js", "Backend", &[r#"require\("#, "node:", r#"process\.env"#]),
        s("express", "Express.js", "Backend", &[r#"from\s+['"]express['"]"#, r#"app\.listen"#]),
        s("django", "Django", "Backend", &[r#"from\s+django"#, "INSTALLED_APPS"]),
        s("flask", "Flask", "Backend", &[r#"from\s+flask"#, r#"@app\.route"#]),
        s("laravel", "Laravel", "Backend", &[r#"use\s+Illuminate"#]),
        s("fastapi", "FastAPI", "Backend", &[r#"from\s+fastapi"#, r#"FastAPI\(\)"#]),
        s("spring-boot", "Spring Boot", "Backend", &["@SpringBoot", "spring-boot"]),

        // ── Mobile ───────────────────────────────────────────────────────────
        s("react-native", "React Native", "Mobile", &["react-native"]),
        s("flutter", "Flutter", "Mobile", &[r#"import\s+['"]package:flutter"#]),
        s("ios", "iOS Development", "Mobile", &[r#"UIKit|SwiftUI|\.xcodeproj"#]),
        s("android", "Android Development", "Mobile", &[r#"android\."#, "AndroidManifest"]),

        // ── Databases ────────────────────────────────────────────────────────
        s("postgresql", "PostgreSQL", "Databases", &[r#"postgres|pg_|pgvector|neon"#, "@neondatabase"]),
        s("mysql", "MySQL", "Databases", &[r#"mysql2?['/]"#]),
        s("mongodb", "MongoDB", "Databases", &["mongodb|mongoose"]),
        s("redis", "Redis", "Databases", &["redis|ioredis"]),
        s("elasticsearch", "Elasticsearch", "Databases", &["elasticsearch|@elastic"]),
        s("cassandra", "Cassandra", "Databases", &["cassandra"]),
        s("dynamodb", "DynamoDB", "Databases", &["dynamodb|DynamoDB"]),
        s("sqlite", "SQLite", "Databases", &["sqlite|better-sqlite"]),
        s("sql", "SQL", "Databases", &[r#"\.sql$"#, r#"SELECT\s+.*FROM"#, r#"INSERT\s+INTO"#]),

        // ── Cloud & DevOps ───────────────────────────────────────────────────
        s("aws", "Amazon Web Services", "Cloud & DevOps", &[r#"@aws-sdk|aws-cdk|s3://|arn:aws"#]),
        s("gcp", "Google Cloud Platform", "Cloud & DevOps", &["@google-cloud|gcloud"]),
        s("azure", "Microsoft Azure", "Cloud & DevOps", &[r#"@azure|azure\."#]),
        s("docker", "Docker", "Cloud & DevOps", &[r#"[Dd]ockerfile|docker-compose|\.dockerignore"#]),
        s("kubernetes", "Kubernetes", "Cloud & DevOps", &["k8s|kubernetes|kubectl|helm"]),
        s("terraform", "Terraform", "Cloud & DevOps", &[r#"\.tf$|terraform"#]),
        s("ansible", "Ansible", "Cloud & DevOps", &[r#"ansible|playbook\.yml"#]),
        s("jenkins", "Jenkins", "Cloud & DevOps", &["[Jj]enkinsfile"]),
        s("ci-cd", "CI/CD", "Cloud & DevOps", &[r#"\.github/workflows|\.gitlab-ci|circleci"#]),
        s("circleci", "CircleCI", "Cloud & DevOps", &[r#"circleci|\.circleci"#]),
        s("serverless", "Serverless", "Cloud & DevOps", &[r#"serverless\.yml|@serverless|edge-runtime"#]),

        // ── Architecture ─────────────────────────────────────────────────────
        s("microservices", "Microservices", "Architecture", &["microservice|service-mesh"]),
        s("rest-api", "REST API", "Architecture", &[r#"app\.(get|post|put|delete|patch)\("#, r#"fetch\(|axios"#]),
        s("graphql", "GraphQL", "Architecture", &[r#"graphql|@apollo|gql|\.graphql$"#, "typeDefs|resolvers"]),
        s("grpc", "gRPC", "Architecture", &[r#"grpc|\.proto$|protobuf"#]),
        s("websocket", "WebSocket", "Architecture", &[r#"[Ww]eb[Ss]ocket|wss?://|ws\.on"#]),
        s("event-driven", "Event-Driven Architecture", "Architecture", &[r#"EventEmitter|pubsub|emit\("#]),

        // ── Tools ────────────────────────────────────────────────────────────
        s("git", "Git", "Tools", &[r#"\.git/|\.gitignore"#]),
        s("linux", "Linux", "Tools", &["/bin/|/usr/|chmod|chown"]),
        s("agile", "Agile", "Tools", &["sprint|kanban|scrum"]),
        s("tdd", "Test-Driven Development", "Tools", &[r#"describe\(|it\(|test\(|expect\(|assert"#]),
        s("webpack", "Webpack", "Tools", &[r#"webpack\.config"#]),
        s("jest", "Jest", "Tools", &[r#"jest\.config"#]),
        s("pytest", "pytest", "Tools", &[r#"import\s+pytest|@pytest"#]),
        s("tailwind", "Tailwind CSS", "Tools", &[r#"tailwind\.config|@tailwind"#]),

        // ── Data Science & ML ────────────────────────────────────────────────
        s("machine-learning", "Machine Learning", "Data Science & ML", &["sklearn|scikit|ml_|machine.?learn"]),
        s("deep-learning", "Deep Learning", "Data Science & ML", &["deep.?learn|neural.?net|backprop"]),
        s("tensorflow", "TensorFlow", "Data Science & ML", &[r#"tensorflow|import\s+tf"#]),
        s("pytorch", "PyTorch", "Data Science & ML", &[r#"import\s+torch|pytorch"#]),
        s("pandas", "Pandas", "Data Science & ML", &[r#"import\s+pandas|pd\.DataFrame"#]),
        s("numpy", "NumPy", "Data Science & ML", &[r#"import\s+numpy|np\.array"#]),
        s("scikit", "scikit-learn", "Data Science & ML", &[r#"from\s+sklearn|scikit"#]),
        s("nlp", "Natural Language Processing", "Data Science & ML", &["nlp|tokeniz|spacy|nltk|huggingface"]),
        s("computer-vision", "Computer Vision", "Data Science & ML", &["computer.?vision|opencv|yolo|image.?classif"]),

        // ── AI / LLM / GenAI ─────────────────────────────────────────────────
        s("llm", "Large Language Models", "AI / LLM", &["llm|large.?language|chat.?completion"]),
        s("rag", "Retrieval-Augmented Generation", "AI / LLM", &[r#"\brag\b|retriev.*generat|retriev.*augment"#]),
        s("prompt-engineering", "Prompt Engineering", "AI / LLM", &["prompt|system.?message|few.?shot"]),
        s("fine-tuning", "Fine-Tuning", "AI / LLM", &["fine.?tun|lora|qlora|adapter"]),
        s("embeddings", "Embeddings", "AI / LLM", &["embed|embedding|vector.?search|cosine.?sim"]),
        s("transformers", "Transformers", "AI / LLM", &["transformer|attention.?head|self.?attention|BertModel"]),
        s("agents", "AI Agents", "AI / LLM", &[r#"\bagent|tool_call|function.?call|agentic"#]),
        s("agentic-ai", "Agentic AI", "AI / LLM", &["agentic|agent.?loop|agent.?team"]),
        s("langchain", "LangChain", "AI / LLM", &[r#"langchain|from\s+langchain"#]),
        s("langgraph", "LangGraph", "AI / LLM", &[r#"langgraph|from\s+langgraph"#]),
        s("openai", "OpenAI", "AI / LLM", &[r#"openai|from\s+['"]openai|gpt-[34]"#]),
        s("anthropic", "Anthropic Claude", "AI / LLM", &["anthropic|claude|@anthropic"]),
        s("vercel-ai-sdk", "Vercel AI SDK", "AI / LLM", &[r#"from\s+['"]ai['"]|from\s+['"]@ai-sdk"#]),
        s("vector-db", "Vector Databases", "AI / LLM", &["vector.?db|lancedb|pgvector|pinecone|weaviate|chromadb"]),
        s("pinecone", "Pinecone", "AI / LLM", &["pinecone|@pinecone"]),
        s("weaviate", "Weaviate", "AI / LLM", &["weaviate"]),
        s("chromadb", "ChromaDB", "AI / LLM", &["chromadb|chroma"]),
        s("mlops", "MLOps", "AI / LLM", &["mlops|mlflow|wandb|weights.?biases"]),
        s("huggingface", "Hugging Face", "AI / LLM", &["huggingface|hf_hub|from_pretrained|transformers"]),
        s("model-evaluation", "Model Evaluation", "AI / LLM", &["eval|benchmark|accuracy|f1.?score|bleu"]),
        s("structured-output", "Structured Output", "AI / LLM", &["structured.?output|json.?schema|zod|response_format"]),
        s("function-calling", "Function Calling", "AI / LLM", &["function.?call|tool.?use|tool_choice"]),
        s("mastra", "Mastra", "AI / LLM", &[r#"mastra|from\s+['"]mastra"#]),

        // ── Cloudflare ───────────────────────────────────────────────────────
        s("cloudflare-workers", "Cloudflare Workers", "Cloudflare", &["cloudflare.*workers|wrangler|miniflare"]),
        s("cloudflare-workers-ai", "Cloudflare Workers AI", "Cloudflare", &["workers.?ai|@cloudflare/ai"]),
        s("cloudflare-d1", "Cloudflare D1", "Cloudflare", &["cloudflare.*d1|D1Database"]),
        s("cloudflare-vectorize", "Cloudflare Vectorize", "Cloudflare", &["vectorize|VectorizeIndex"]),

        // ── Frontend (extended) ──────────────────────────────────────────────
        s("next-auth", "NextAuth.js", "Frontend", &["next-auth|NextAuth"]),
        s("radix-ui", "Radix UI", "Frontend", &["@radix-ui"]),
        s("shadcn-ui", "shadcn/ui", "Frontend", &["shadcn|@/components/ui"]),
        s("storybook", "Storybook", "Frontend", &[r#"storybook|\.stories\."#]),
        s("playwright", "Playwright", "Frontend", &["playwright|@playwright"]),
        s("cypress", "Cypress", "Frontend", &[r#"cypress|cy\."#]),
        s("vitest", "Vitest", "Frontend", &[r#"vitest|from\s+['"]vitest"#]),
        s("react-query", "React Query / TanStack Query", "Frontend", &["@tanstack/react-query|useQuery|useMutation"]),
        s("zustand", "Zustand", "Frontend", &[r#"zustand|create\(\s*\(set"#]),
        s("apollo-client", "Apollo Client", "Frontend", &["@apollo/client|useQuery|useMutation|ApolloProvider"]),
        s("remix", "Remix", "Frontend", &[r#"from\s+['"]@remix|remix\.config"#]),
        s("astro", "Astro", "Frontend", &[r#"\.astro$|astro\.config"#]),

        // ── Backend (extended) ───────────────────────────────────────────────
        s("drizzle-orm", "Drizzle ORM", "Backend", &[r#"drizzle-orm|drizzle-kit|from\s+['"]drizzle"#]),
        s("prisma", "Prisma", "Backend", &[r#"@prisma|prisma\.schema"#]),
        s("trpc", "tRPC", "Backend", &["@trpc|trpc"]),
        s("hono", "Hono", "Backend", &[r#"from\s+['"]hono"#]),
        s("bun", "Bun", "Backend", &[r#"bun\.sh|bunfig|Bun\.serve"#]),
        s("deno", "Deno", "Backend", &[r#"deno\.|Deno\."#]),
    ];

    entries.into_iter().map(|s| (s.tag, s)).collect()
});

fn s(tag: &'static str, label: &'static str, category: &'static str, patterns: &'static [&'static str]) -> SkillDef {
    SkillDef { tag, label, category, patterns }
}

/// Look up a taxonomy entry by slug.
pub fn get(tag: &str) -> Option<&'static SkillDef> {
    TAXONOMY.get(tag)
}

/// All category names present in the taxonomy.
pub fn categories() -> Vec<&'static str> {
    let mut cats: Vec<&str> = TAXONOMY.values().map(|s| s.category).collect();
    cats.sort();
    cats.dedup();
    cats
}
